### DISCUSSION TITLE
Error with Webpack + Prisma

### DISCUSSION THREAD

**Discussion by Bradleykingz** [prisma/prisma#3161] [Q&A] [answered]
I'm trying to use webpack to bundle my code but I keep getting the following error:

```
Error (f) {
message: `Query engine binary for current platform "debian-openssl-1.1.x" could not be found.␊
This probably happens, because you built Prisma Client on a different platform.␊
(Prisma Client looked in "/query-engine-debian-openssl-1.1.x")␊
  
Files in /:␊
    ␊
      bin␊
      boot␊
      cdrom␊
      data␊
      dev␊
      etc␊
      home␊
//...
```
Now, I've run into this error before, but I fixed it by copying over the prisma folder to the 'build' directory. This time, I simply can't figure out why it would try to look inside the actual **root** folder. Is this something that can be configured? Or what could I be doing wrong?

**Comment by Bradleykingz** [comment]
I've enabled debug mode and here is some of the output

```log
2020-07-29T08:49:48.652Z engine { cwd: '/' }
2020-07-29T08:49:48.678Z engine { dotPrismaPath: '/.prisma/client/query-engine-debian-openssl-1.1.x' }
2020-07-29T08:49:48.678Z engine { dirnamePath: '/query-engine-debian-openssl-1.1.x' }
2020-07-29T08:49:48.678Z engine { parentDirName: '/query-engine-debian-openssl-1.1.x' }
2020-07-29T08:49:48.679Z engine { prismaPath: '/query-engine-debian-openssl-1.1.x' }
```

It seems like the root directory is being set as the current working directory at some point. Should that normally happen?

**Reply by ryands17** [reply]
Hey @Bradleykingz 👋 
Could you share your webpack config. I would like to check this at my end as well.

**Reply by Bradleykingz** [reply]
Here it is.

```js
const path = require('path');
const nodeExternals = require('webpack-node-externals');
const {resolveTsAliases} = require("resolve-ts-aliases");
const WebpackShellPlugin = require('webpack-shell-plugin');
const CircularDependencyPlugin = require('circular-dependency-plugin');
const CopyWebpackPlugin = require('copy-webpack-plugin');

module.exports = {
    entry: {
        app: './webpack/all-test.js' // only bundle test files + dependencies
    },
    target: "node",
    mode: "production",
    name: "ci",
    module: {
        rules: [
            {
                test: /\.ts?$/,
                use: 'ts-loader',
                include: path.resolve(__dirname, 'src'),
            },
        ],
    },
    resolve: {
        alias: resolveTsAliases(path.resolve("tsconfig.json")),
        extensions: ['.ts', '.js']
    },
    output: {
        path: path.resolve(__dirname, 'build/bundle'),
        filename: "[name].suite.test.js"
    },
    externals: [nodeExternals()],
    plugins: [
        new CopyWebpackPlugin({
           patterns: [
               {
                   from: 'src/db/prisma',
                   to: 'db/prisma', // this is called within the 'build/bundle' folder.
                   globOptions: {
                       ignore: ['**/*.d.ts'],
                   },
               }
           ],

        }),
        new WebpackShellPlugin({
            onBuildExit: [
                'npm run test' // error happens at this point
            ]
        })
    ]
}
```

```js
// all-test.js
const context = require.context('../src/tests', true, /.ts$/);
context.keys().forEach(context);
module.exports = context;
```

**Reply by ryands17** [reply]
Are you bundling the application to a serverless environment? If not, then you do not need to copy any files to any directory as Prisma uses the root directory so whatever Webpack outputs, should work without any configuration.

You have tests included, as an example I have created an Express app that works fine after building by Webpack without any additional copying of any files.

**Reply by Bradleykingz** [reply]
Do you mean it uses the  **root project** directory directory? Because my problem is that Prisma keeps looking in the **root linux** directory, where it won't ever find anything.

Bundling and running it without copying results in the same error, and copying the prisma folder to the root of my project doesn't seem to do anything, either.

**Reply by Bradleykingz** [reply]
Maybe to help my case, are there situations where Prisma is expected to use the root directory like that? Maybe I could work backwards and track down my mistake.

Otherwise, where does it normally expect to find the query engine?

**Reply by ryands17** [reply]
The basic folder structure that I have scaffolded is as follows:
![image](https://user-images.githubusercontent.com/19697099/88794029-6b72da00-d1bb-11ea-9429-04af8c716eb8.png)

By default, Prisma looks in your root directory inside the `prisma` folder for the `schema.prisma` file.
What's exactly in the `src/db/prisma` folder?

**Reply by Bradleykingz** [reply]
My project is written in typescript and has two top-level directories - build and src.

`src/db/prisma` is where generated prisma files are stored.

```
src/db/
├── dao
│   ├── ...models
├── prisma
│   ├── generated
│   │   └── client
│   │       ├── index.js
│   │       ├── package.json
│   │       ├── query-engine-debian-openssl-1.1.x
│   │       ├── runtime
│   │       │   ├── highlight
│   │       │   │   └── languages
│   │       │   ├── index.js
│   │       │   ├── index.js.map
│   │       │   └── utils
│   │       └── schema.prisma
│   └── schema.prisma
└── sql
    ├── ...sql files

```

**Reply by ryands17** [reply]
So you're storing the generated output in a custom directory? If not, it would be great if you could share your code so that I can check how to fix that for custom directories.

**Reply by Bradleykingz** [reply]
Yeah. I'm storing the generated output inside `src/db/prisma/generated/client`.

Should I share my `schema.prisma`?

**Reply by ryands17** [reply]
Yes that would be great!

**Reply by Bradleykingz** [reply]
There's not much to it, really.

```prisma
generator client {
  provider      = "prisma-client-js"
  output        = "generated/client"
  binaryTargets = ["native"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

// the rest is generated output
```

**Reply by ryands17** [reply]
Could you try running `prisma generate` after copying the files and check again?

**Reply by Bradleykingz** [reply]
Oh, you wanted the whole file? Can I send it via email or something? It's both large and I'm not so comfortable posting it in public.

**Reply by ryands17** [reply]
No it's fine. I'll reproduce with the custom directory and let you know.

**Reply by Bradleykingz** [reply]
Sure. Thanks!

**Reply by Bradleykingz** [reply]
Btw, I can confirm that generating prisma inside `node_modules` works fine.

**Comment by ryands17** [comment, marked as answer]
[webpack.zip](https://github.com/prisma/prisma/files/5000361/webpack.zip)

This is my example with a custom directory for generate. I couldn't reproduce it with this setup. Could you try changeing/updating this setup as per you structure or try with this webpack config?

**Reply by Bradleykingz** [reply]
I managed to reproduce it.

[prisma-webpack.tar.gz](https://github.com/prisma/prisma/files/5001447/prisma-webpack.tar.gz)

**Reply by ryands17** [reply]
I think this is an issue with how Webpack handles`__dirname`. The path is being set to the `dist` folder directly which is causing issues. One option is to exclude the Prisma Client JS file from the bundle and use the `CopyWebpackPlugin` to copy the entire directory.
I will try check for a solution, but if possible I would recommend avoiding Webpack for any backend related setup.

**Reply by Bradleykingz** [reply]
I've tried messing with `IgnorePlugin` and `rule.exclude` from webpack with no luck. 

I'll just proceed with the default folder for now.

Thanks for all the help!

**Reply by ryands17** [reply]
Just a side question though, what was the exact use-case of integrating Webpack here?

**Reply by Bradleykingz** [reply]
I previously used babel-node to run tests but waiting for every file to transpile before running a test is **slow.**

Instead, I use webpack in the CI stage. The test files are bundled into a single output and run pretty fast. I've also found that running Webpack with babel is much faster than `ts > js` (building) then `js > js` (transpiling) when using something like gulp.

**Reply by manavm1990** [reply]
Given this architecture:
![image](https://user-images.githubusercontent.com/3387791/94827866-884fa780-03ce-11eb-8bd7-97df212f02c7.png)

All that I had to do to eliminate all kinds of 'yellow terminal warnings' and to get this 💩 to work was to have this in 'webpack.config.js'

`const nodeExternals = require("webpack-node-externals");` and `externals: [nodeExternals()],` in `module.exports`.

I am still not 💯  sure how/why this magically gets `prisma` to work properly without being super 😕 about where to find what it needs ('query engine') but I'll take it.

I've never used this particular module b4. So. many 'hidden things' with 'webpack.config.js' 😫 .

On a funny aside - when it actually built without warnings, I was thrown off that I thought my system was just hanging - I shut it down and redid it a couple of times. I had gotten so used to seeing all of the `webpack` warnings and had accepted that as 'normal'.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
